File: TA/Backend/SoftRhRMI_Servidor/src/pe/edu/pucp/softrh/rmi/interfaces/ClienteBO.java
package pe.edu.pucp.softrh.rmi.interfaces;

import java.rmi.Remote;
import java.rmi.RemoteException;
import java.util.ArrayList;
import java.util.Date;
import pe.edu.pucp.softrh.usuarios.model.Cliente;

public interface ClienteBO extends Remote{
    Integer insertar(String dni, String nombres, String apellidos, String correo, String contrasenha, String telefono, Date fechaRegistro, Boolean recibePromociones) throws RemoteException;

    Integer modificar(Integer idUsuario, String dni, String nombres, String apellidos, String correo, String telefono, Date fechaRegistro, Boolean recibePromociones) throws RemoteException;

    Integer eliminar(Integer idCliente) throws RemoteException;

    ArrayList<Cliente> listarTodos() throws RemoteException;

    Cliente obtenerPorId(Integer idCliente) throws RemoteException;

    ArrayList<Cliente> listarPorDniNombre(String cadena) throws RemoteException;

    Integer verificarCorreoExistente(String correo) throws RemoteException;
}

File: TA/Frontend/RHStoreWS/RHStoreUsuariosBO/ClienteBO.cs
using RHStoreBaseBO;
using RHStoreBaseBO.ServiciosWeb;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RHStoreUsuariosBO
{
	public class ClienteBO : BaseBO
	{
        public int insertar(string dni, string nombres, string apellidos, string correo, string contrasenha, string telefono, DateTime fechaRegistro, bool recibePromociones)
        {
            return ClienteWS.insertarCliente(dni, nombres, apellidos, correo, contrasenha, telefono, fechaRegistro, recibePromociones);
        }

        public int modificar(int idUsuario, string dni, string nombres, string apellidos, string correo, string telefono, DateTime fechaRegistro, bool recibePromociones)
        {
            return ClienteWS.modificarCliente(idUsuario, dni, nombres, apellidos, correo, telefono, fechaRegistro, recibePromociones);
        }

        public int eliminar(int idCliente)
        {
            return ClienteWS.eliminarCliente(idCliente);
        }

        public BindingList<cliente> listarTodos()
        {
  
            try
            {
                List<cliente> clientes = ClienteWS.listarClientesTodos()?.ToList();
                return new BindingList<cliente>(clientes ?? new List<cliente>());
            }
            catch (Exception ex)
            {
                // Manejar la excepción (registrarla o notificar al usuario)
                return new BindingList<cliente>(new List<cliente>());
            }

        }

        public cliente obtenerPorId(int idCliente)
        {
            return ClienteWS.obtenerClientePorId(idCliente);
        }

		public BindingList<cliente> listarPorDniNombre(string cadena)
		{
            try
            {
                List<cliente> clientes = ClienteWS.listarClientesPorDniNombre(cadena)?.ToList();
                return new BindingList<cliente>(clientes ?? new List<cliente>());
            }
            catch (Exception ex)
            {
                // Manejar la excepción (registrarla o notificar al usuario)
                return new BindingList<cliente>(new List<cliente>());
            }

        }

		public int verificarCorreoExistente(string correo)
		{
			return ClienteWS.verificarCorreoExistente(correo);
		}
	}
}
